core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -247,18 +247,18 @@ private[spark] class MemoryStore(
         val amountToRequest = size - unrollMemoryUsedByThisBlock
         keepUnrolling = reserveUnrollMemoryForThisTask(blockId, amountToRequest, memoryMode)
         if (keepUnrolling) {
-          unrollMemoryUsedByThisBlock += amountToRequest
+          unrollMemoryUsedByThisBlock = size
         }
+      } else if (size < unrollMemoryUsedByThisBlock) {
+        releaseUnrollMemoryForThisTask(memoryMode, unrollMemoryUsedByThisBlock - size)
+        unrollMemoryUsedByThisBlock = size
       }
 
       if (keepUnrolling) {
         val entry = entryBuilder.build()
-        // Synchronize so that transfer is atomic
-        memoryManager.synchronized {
-          releaseUnrollMemoryForThisTask(memoryMode, unrollMemoryUsedByThisBlock)
-          val success = memoryManager.acquireStorageMemory(blockId, entry.size, memoryMode)
-          assert(success, "transferring unroll memory to storage memory failed")
-        }
+        // In fact, unroll memory is also storage memory, it is unnecessary to
+        // release unroll memory really, and then to get storage memory again.
+        releaseUnrollMemoryForThisTask(memoryMode, unrollMemoryUsedByThisBlock, false)
```

Comment:
You add a parameter 'releaseMemoryReally' here, so that memory will not release unroll memory when it is false.
If it is false, will not release storage memory but still decrease unrollMemoryMap.
This improvment looks like avoid repeated release/acquire on storage memory.

Reply:
Yeah,thanks